File: lib/ghost.ex
defmodule Ghost do
  use HTTPotion.Base
  alias Ghost.Config

  def process_url(url), do: "https://api.github.com" <> url
  def process_request_headers(hdrs), do: Dict.put(hdrs, :"User-Agent", "ghost")

  def process_response_body(body) do
    body |> to_string |> :jsx.decode
  end

  def opts() do
    [ibrowse: [{:basic_auth, 
        {Config.get[:tok] |> List.from_char_data!, 'x-oauth-basic'}}]]
  end

  def get(url), do: get(url, [], opts)

  def gists(filter_row \\ &id/1, filter_col \\ &sel/1) do
    get("/users/#{Config.get[:me]}/gists").body
    |> Enum.filter(filter_row)
    |> Enum.map(filter_col)
  end

  defp id(x), do: x

  def sel(gist, fields \\ ["id", "files", "public"]) do
    fields = fields |> Enum.into HashSet.new
    gist |> Dict.to_list |> Enum.filter(fn {k, _v} ->
      Set.member?(fields, k)
    end)
  end

end

File: lib/ghost.rb
require 'faraday'
require 'json'

module Ghost
  module_function
  def conn()
    @conn ||= Faraday.new(url: "https://api.github.com") do |client|
      #client.response :logger
      client.adapter Faraday.default_adapter
    end.tap do |conn|
      @config = Hash.new.tap do |h|
        File.open(ENV['HOME']+'/.ghostrc') do |f| 
        while(s = f.gets) do
          k, v = process_line(s)
          h[k] = v
        end; end
      end
      conn.basic_auth @config['tok'], 'x-oauth-basic'
    end
  end

  def get(*args) JSON.parse(conn.get(*args).body) end

  def config() @config end

  def gists()
    get("/users/#{@config['me']}/gists")
  end
 
  def del() @conn=nil end #tmp

  def process_line(line)
    line.chomp.split('=').map &:strip
  end
end
